File: pkg/@eser/cs/sync.ts
// Copyright 2023-present Eser Ozvataf and other contributors. All rights reserved. Apache-2.0 license.

import { runtime } from "@eser/standards/runtime";
import * as writer from "@eser/writer";
import * as dotenv from "@eser/config/dotenv";
import type {
  ConfigMap,
  KubectlResourceReference,
  Secret,
  SyncOptions,
} from "./types.ts";

// Constants
const K8S_MAX_NAME_LENGTH = 253; // RFC 1123 DNS subdomain max length
const ERROR_PREVIEW_LENGTH = 100; // Characters to show in error previews

/**
 * Kubernetes resource name validation error
 */
export class KubernetesResourceNameError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "KubernetesResourceNameError";
  }
}

/**
 * Validates a Kubernetes resource name according to RFC 1123 DNS subdomain naming rules.
 * - Must be 253 characters or less
 * - Must contain only lowercase alphanumeric characters, '-', or '.'
 * - Must start and end with an alphanumeric character
 *
 * @param name - The resource name to validate
 * @param fieldName - The field name for error messages (e.g., "resource name", "namespace")
 * @throws {KubernetesResourceNameError} If the name is invalid
 */
export const validateKubernetesResourceName = (
  name: string,
  fieldName: string = "resource name",
): void => {
  if (name === undefined || name === null) {
    throw new KubernetesResourceNameError(`${fieldName} is required`);
  }

  if (typeof name !== "string") {
    throw new KubernetesResourceNameError(`${fieldName} must be a string`);
  }

  if (name.length === 0) {
    throw new KubernetesResourceNameError(`${fieldName} cannot be empty`);
  }

  if (name.length > K8S_MAX_NAME_LENGTH) {
    throw new KubernetesResourceNameError(
      `${fieldName} must be 253 characters or less (got ${name.length})`,
    );
  }

  // RFC 1123 DNS subdomain: lowercase alphanumeric, '-', or '.'
  // Must start and end with alphanumeric
  const validPattern = /^[a-z0-9]([a-z0-9.-]*[a-z0-9])?$/;

  if (!validPattern.test(name)) {
    throw new KubernetesResourceNameError(
      `${fieldName} "${name}" is invalid. Must contain only lowercase alphanumeric characters, '-', or '.', and must start and end with an alphanumeric character`,
    );
  }

  // Check for consecutive dots or dashes which are not allowed
  if (name.includes("..") || name.includes("--")) {
    throw new KubernetesResourceNameError(
      `${fieldName} "${name}" cannot contain consecutive dots or dashes`,
    );
  }
};

/**
 * Validates a KubectlResourceReference, checking both name and namespace if provided.
 *
 * @param resource - The resource reference to validate
 * @throws {KubernetesResourceNameError} If validation fails
 */
export const validateResourceReference = (
  resource: KubectlResourceReference,
): void => {
  validateKubernetesResourceName(resource.name, "resource name");

  if (resource.namespace !== undefined && resource.namespace !== null) {
    validateKubernetesResourceName(resource.namespace, "namespace");
  }
};

/**
 * Executes a kubectl get command to retrieve keys from a Kubernetes resource.
 * Uses direct command invocation (no shell) for security.
 *
 * @param resource - The Kubernetes resource reference
 * @returns Array of data keys from the resource
 * @throws {Error} If kubectl command fails
 * @throws {KubernetesResourceNameError} If resource name/namespace is invalid
 */
export const executeKubectl = async (
  resource: KubectlResourceReference,
): Promise<string[]> => {
  // Validate inputs before executing any commands
  validateResourceReference(resource);

  const resourceType = resource.type === "configmap" ? "cm" : "secret";

  // Build command arguments array (safe from injection)
  const args: string[] = [
    "get",
    `${resourceType}/${resource.name}`,
  ];

  if (resource.namespace) {
    args.push("-n", resource.namespace);
  }

  args.push("-o", "json");

  // Execute kubectl directly without shell wrapper
  const { success, stdout, stderr } = await runtime.exec.spawn(
    "kubectl",
    args,
    {
      stdout: "piped",
      stderr: "piped",
    },
  );

  if (!success) {
    const errorMessage = new TextDecoder().decode(stderr);
    throw new Error(`kubectl command failed: ${errorMessage}`);
  }

  // Parse JSON in JavaScript instead of using jq (removes external dependency)
  const output = new TextDecoder().decode(stdout);

  try {
    const resourceData = JSON.parse(output);
    const data = resourceData.data ?? {};
    return Object.keys(data);
  } catch (_parseError) {
    throw new Error(
      `Failed to parse kubectl output as JSON: ${
        output.substring(0, ERROR_PREVIEW_LENGTH)
      }...`,
    );
  }
};

/**
 * Builds a Kubernetes Secret resource from the given context.
 *
 * @param name - The Secret name
 * @param namespace - The namespace (optional)
 * @param data - Map of key-value pairs to include in the Secret
 * @returns A Kubernetes Secret resource object
 *
 * @remarks
 * **Security Note:** The btoa() function performs Base64 encoding, which is NOT encryption.
 * Kubernetes Secrets are stored unencrypted in etcd by default. Base64 encoding is used
 * to support binary data and is required by the Kubernetes API, but it provides no
 * confidentiality. For sensitive data, consider:
 * - Enabling etcd encryption at rest
 * - Using external secret management (HashiCorp Vault, AWS Secrets Manager, etc.)
 * - Using sealed-secrets or similar solutions
 */
export const buildSecretFromContext = (
  name: string,
  namespace: string | undefined,
  data: dotenv.EnvMap,
): Secret => {
  const encodedData: Record<string, string> = {};

  // Base64 encode data for Kubernetes API compatibility (NOT for security)
  for (const [key, value] of data) {
    encodedData[String(key)] = btoa(value);
  }

  return {
    apiVersion: "v1",
    kind: "Secret",
    metadata: {
      name,
      ...(namespace !== undefined && namespace !== "default"
        ? { namespace }
        : {}),
    },
    data: encodedData,
    type: "Opaque",
  };
};

export const buildConfigMapFromContext = (
  name: string,
  namespace: string | undefined,
  data: dotenv.EnvMap,
): ConfigMap => {
  return {
    apiVersion: "v1",
    kind: "ConfigMap",
    metadata: {
      name,
      ...(namespace && namespace !== "default" ? { namespace } : {}),
    },
    data: Object.fromEntries(data.entries()),
  };
};

/**
 * Synchronizes environment variables with a Kubernetes ConfigMap or Secret.
 * Generates a kubectl patch command that can be executed to update the resource.
 *
 * @param options - Sync configuration options
 * @returns A kubectl patch command string or the patch data (if stringOnly is true)
 * @throws {KubernetesResourceNameError} If resource name/namespace is invalid
 * @throws {Error} If kubectl command fails or sync operation fails
 *
 * @example
 * ```typescript
 * const command = await sync({
 *   resource: { type: "configmap", name: "my-config", namespace: "default" },
 *   format: "json"
 * });
 * // Returns: kubectl patch cm my-config -n default --type=merge -p '{"data":{...}}'
 * ```
 */
export const sync = async (options: SyncOptions): Promise<string> => {
  // Register formats lazily when needed
  writer.registerBuiltinFormats();

  try {
    // Validate resource reference early (defense in depth)
    // executeKubectl also validates, but we validate here for early failure
    validateResourceReference(options.resource);

    // Get keys from Kubernetes resource
    const keys = await executeKubectl(options.resource);

    if (keys.length === 0) {
      return `# No data found in ${options.resource.type}/${options.resource.name}`;
    }

    // Load environment data using @eser/config
    const envMap = await dotenv.load({ env: options.env });

    // Convert Map to Record and filter for the keys we need
    const envData: Record<string, string> = {};
    for (const key of keys) {
      const value = envMap.get(key);
      if (value !== undefined) {
        envData[key] = value;
      } else {
        console.warn(
          `Warning: Environment variable ${key} not found in environment`,
        );
      }
    }

    if (Object.keys(envData).length === 0) {
      return `# No matching environment variables found for ${options.resource.type}/${options.resource.name}`;
    }

    // Prepare data for patch
    let patchData: Record<string, string>;

    if (options.resource.type === "secret") {
      // Base64 encode values for Kubernetes API (NOT for security - see buildSecretFromContext docs)
      patchData = {};
      for (const [key, value] of Object.entries(envData)) {
        patchData[key] = btoa(value);
      }
    } else {
      // Use values as-is for configmaps
      patchData = envData;
    }

    // Generate patch content based on format
    const format = options.format ?? "json";
    const patchObject = { data: patchData };

    let patchString: string;
    if (format === "yaml") {
      patchString = writer.write([patchObject], "yaml", { pretty: true })
        .trim();
    } else {
      patchString = JSON.stringify(patchObject);
    }

    // Return string-only if requested
    if (options.stringOnly) {
      return patchString;
    }

    // Generate kubectl patch command
    // Resource name and namespace have been validated above (RFC 1123 compliant)
    // so they are safe to include in the command string
    const resourceType = options.resource.type === "configmap"
      ? "cm"
      : "secret";
    const namespaceFlag = options.resource.namespace
      ? ` -n ${options.resource.namespace}`
      : "";

    // Escape single quotes in the patch data for shell usage
    const escapedPatch = patchString.replace(/'/g, "'\"'\"'");

    return `kubectl patch ${resourceType} ${options.resource.name}${namespaceFlag} --type=merge -p '${escapedPatch}'`;
  } catch (error) {
    if (error instanceof KubernetesResourceNameError) {
      throw error; // Re-throw validation errors as-is
    }
    throw new Error(
      `Failed to sync with kubectl: ${
        error instanceof Error ? error.message : String(error)
      }`,
    );
  }
};

/**
 * Generates a command to apply the sync patch using kubectl apply.
 *
 * @param options - Sync configuration options
 * @returns A shell command string that applies the patch
 * @throws {KubernetesResourceNameError} If resource name/namespace is invalid
 * @throws {Error} If sync operation fails
 *
 * @remarks
 * This function generates a shell command string. The resource name and namespace
 * are validated according to RFC 1123 DNS subdomain naming rules before being
 * included in the output.
 */
export const syncApply = async (options: SyncOptions): Promise<string> => {
  // Validate early for better error messages
  validateResourceReference(options.resource);

  const manifest = await sync({ ...options, stringOnly: true });

  if (manifest.startsWith("# No")) {
    return manifest;
  }

  // Generate a heredoc-style command which is safer than echo with escaping
  // Resource name/namespace are already validated by sync()
  const resourceType = options.resource.type === "configmap" ? "cm" : "secret";
  const namespaceFlag = options.resource.namespace
    ? ` -n ${options.resource.namespace}`
    : "";

  // Use heredoc for safer shell command generation
  return `kubectl patch ${resourceType} ${options.resource.name}${namespaceFlag} --type=merge -p '${
    manifest.replace(/'/g, "'\"'\"'")
  }'`;
};

File: pkg/ajan/csfx/sync.go
// Copyright 2023-present Eser Ozvataf and other contributors. All rights reserved. Apache-2.0 license.

package csfx

import (
	"encoding/base64"
	"encoding/json"
	"errors"
	"fmt"
	"os/exec"
	"strings"

	"github.com/eser/stack/pkg/ajan/formatfx"
)

// RunKubectl retrieves the data keys from an existing Kubernetes resource by
// executing `kubectl get <type>/<name> -o json`. Returns an error when the
// kubectl command fails or its JSON output cannot be parsed.
func RunKubectl(ref ResourceReference) ([]string, error) {
	if err := ValidateResourceReference(ref); err != nil {
		return nil, err
	}

	resourceType := "cm"
	if ref.Type == ResourceTypeSecret {
		resourceType = "secret"
	}

	args := []string{"get", resourceType + "/" + ref.Name}

	if ref.Namespace != "" {
		args = append(args, "-n", ref.Namespace)
	}

	args = append(args, "-o", "json")

	output, err := exec.Command("kubectl", args...).Output() //nolint:gosec // arguments validated above

	if err != nil {
		var exitErr *exec.ExitError
		if errors.As(err, &exitErr) {
			return nil, fmt.Errorf("kubectl command failed: %s", strings.TrimSpace(string(exitErr.Stderr)))
		}

		return nil, fmt.Errorf("kubectl command failed: %w", err)
	}

	var resourceJSON struct {
		Data map[string]any `json:"data"`
	}

	if err := json.Unmarshal(output, &resourceJSON); err != nil {
		preview := string(output)
		if len(preview) > 100 { //nolint:mnd
			preview = preview[:100] + "..."
		}

		return nil, fmt.Errorf("failed to parse kubectl output as JSON: %w (output: %s)", err, preview)
	}

	keys := make([]string, 0, len(resourceJSON.Data))

	for k := range resourceJSON.Data {
		keys = append(keys, k)
	}

	return keys, nil
}

// Sync generates a kubectl patch command that updates an existing Kubernetes
// resource with values from a local .env file, filtered to only the keys
// already present in the resource (obtained via RunKubectl).
//
// When opts.StringOnly is true, only the raw patch data string is returned
// (no kubectl command wrapper).
func Sync(opts SyncOptions) (string, error) {
	formatfx.RegisterBuiltinFormats()

	if err := ValidateResourceReference(opts.Resource); err != nil {
		return "", err
	}

	keys, err := RunKubectl(opts.Resource)
	if err != nil {
		return "", fmt.Errorf("sync: %w", err)
	}

	if len(keys) == 0 {
		return fmt.Sprintf("# No data found in %s/%s", opts.Resource.Type, opts.Resource.Name), nil
	}

	envData, err := loadEnvFile(opts.EnvFile)
	if err != nil {
		return "", err
	}

	// Filter env data to only keys present in the K8s resource.
	filtered := make(map[string]string, len(keys))

	for _, k := range keys {
		if v, ok := envData[k]; ok {
			filtered[k] = v
		}
	}

	if len(filtered) == 0 {
		return fmt.Sprintf("# No matching environment variables found for %s/%s", opts.Resource.Type, opts.Resource.Name), nil
	}

	// Secrets require Base64-encoded values in the data field.
	patchData := filtered

	if opts.Resource.Type == ResourceTypeSecret {
		patchData = make(map[string]string, len(filtered))

		for k, v := range filtered {
			patchData[k] = base64.StdEncoding.EncodeToString([]byte(v))
		}
	}

	format := opts.Format
	if format == "" {
		format = "json"
	}

	patchObject := map[string]any{"data": patchData}

	var patchString string

	switch format {
	case "yaml":
		f, fErr := formatfx.GetFormat("yaml")
		if fErr != nil {
			return "", fmt.Errorf("sync: %w", fErr)
		}

		s, wErr := f.WriteItem(patchObject, &formatfx.FormatOptions{Pretty: true}) //nolint:exhaustruct
		if wErr != nil {
			return "", fmt.Errorf("sync: %w", wErr)
		}

		patchString = strings.TrimSpace(s)

	default:
		b, mErr := json.Marshal(patchObject)
		if mErr != nil {
			return "", fmt.Errorf("sync: %w", mErr)
		}

		patchString = string(b)
	}

	if opts.StringOnly {
		return patchString, nil
	}

	return kubectlPatchCommand(opts.Resource, patchString), nil
}

// SyncApply is a convenience wrapper that validates the resource reference and
// then returns a kubectl patch command. It is equivalent to Sync with
// StringOnly=false.
func SyncApply(opts SyncOptions) (string, error) {
	if err := ValidateResourceReference(opts.Resource); err != nil {
		return "", err
	}

	return Sync(SyncOptions{
		Resource:   opts.Resource,
		EnvFile:    opts.EnvFile,
		Format:     opts.Format,
		StringOnly: false,
	})
}

// kubectlPatchCommand builds a `kubectl patch` command string.
// The resource name and namespace are validated before reaching this point so
// they are safe to embed in the command string.
func kubectlPatchCommand(ref ResourceReference, patchString string) string {
	resourceType := "cm"
	if ref.Type == ResourceTypeSecret {
		resourceType = "secret"
	}

	var sb strings.Builder

	sb.WriteString("kubectl patch ")
	sb.WriteString(resourceType)
	sb.WriteByte(' ')
	sb.WriteString(ref.Name)

	if ref.Namespace != "" {
		sb.WriteString(" -n ")
		sb.WriteString(ref.Namespace)
	}

	sb.WriteString(" --type=merge -p '")
	// Escape single-quotes for POSIX shell safety.
	sb.WriteString(strings.ReplaceAll(patchString, "'", "'\"'\"'"))
	sb.WriteByte('\'')

	return sb.String()
}
